app/api/groups/get_user_progress.go

```diff
@@ -0,0 +1,269 @@
+				SELECT ID FROM groups_attempts
+				WHERE idGroup = team_links.idGroupParent AND idItem = items.ID AND sLastActivityDate IS NOT NULL
+				ORDER BY sLastActivityDate DESC LIMIT 1
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS last_attempt
+			ON last_attempt.ID = IF(
+				(
+					last_attempt_of_team.ID IS NOT NULL AND
+					last_attempt_of_user.ID IS NOT NULL AND
+					last_attempt_of_team.sLastActivityDate > last_attempt_of_user.sLastActivityDate
+				) OR last_attempt_of_user.ID IS NULL,
+				last_attempt_of_team.ID,
+				last_attempt_of_user.ID
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_attempt_of_user
+			ON first_attempt_of_user.ID = (
+				SELECT ID FROM groups_attempts
+				WHERE idGroup = groups.ID AND idItem = items.ID AND sStartDate IS NOT NULL
+				ORDER BY sStartDate LIMIT 1
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_attempt_of_team
+			ON first_attempt_of_team.ID = (
+				SELECT ID FROM groups_attempts
+				WHERE idGroup = team_links.idGroupParent AND idItem = items.ID AND sStartDate IS NOT NULL
+				ORDER BY sStartDate LIMIT 1
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_attempt
+			ON first_attempt.ID = IF(
+				(
+					first_attempt_of_team.ID IS NOT NULL AND
+					first_attempt_of_user.ID IS NOT NULL AND
+					first_attempt_of_team.sStartDate < first_attempt_of_user.sStartDate
+				) OR first_attempt_of_user.ID IS NULL,
+				first_attempt_of_team.ID,
+				first_attempt_of_user.ID
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_validated_attempt_of_user
+			ON first_validated_attempt_of_user.ID = (
+				SELECT ID FROM groups_attempts
+				WHERE idGroup = groups.ID AND idItem = items.ID AND sValidationDate IS NOT NULL
+				ORDER BY sValidationDate LIMIT 1
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_validated_attempt_of_team
+			ON first_validated_attempt_of_team.ID = (
+				SELECT ID FROM groups_attempts
+				WHERE idGroup = team_links.idGroupParent AND idItem = items.ID AND sValidationDate IS NOT NULL
+				ORDER BY sValidationDate LIMIT 1
+			)`).
+		Joins(`
+			LEFT JOIN groups_attempts AS first_validated_attempt
+			ON first_validated_attempt.ID = IF(
+				(
+					first_validated_attempt_of_team.ID IS NOT NULL AND
+					first_validated_attempt_of_user.ID IS NOT NULL AND
+					first_validated_attempt_of_team.sValidationDate < first_validated_attempt_of_user.sValidationDate
+				) OR first_attempt_of_user.ID IS NULL,
+				first_validated_attempt_of_team.ID,
+				first_validated_attempt_of_user.ID
+			)`).
+		Where("groups.ID IN (?)", userGroupIDs).
+		Group("groups.ID, items.ID").
+		Order(gorm.Expr(
+			"FIELD(groups.ID"+strings.Repeat(", ?", len(userGroupIDs))+")",
+			userGroupIDs...)).
+		Order(gorm.Expr(
+			"FIELD(items.ID"+strings.Repeat(", ?", len(itemIDs))+")",
+			itemIDs...)).
+		Order("attempt_with_best_score.iMinusScore, attempt_with_best_score.sBestAnswerDate").
+		ScanIntoSliceOfMaps(&dbResult).Error())
```

Comment:
(cont. of what I wrote above)
Globally, I would have expected just take all the attempts from the group_id IN (user self id or all teams the user belongs to), and just do the max etc on this. Would prevent the 3 joins for each operations, would be much much simpler... but maybe that's less efficient (?).
Or I missed something 🙂

Reply:
That's exactly how I implemented the task in the beginning. The problem is that it doesn't work on large data sets (it didn't even finish in 15,000 seconds): ```SELECT ID FROM groups_attempts WHERE idGroup IN (...) AND idItem = ... ORDER BY idItem, iMinusScore, sBestAnswerDate LIMIT 1``` doesn't use indices at all. That's why I finally implemented the task this way. Yes it looks strange, but each of these three JOINs uses the index, so the whole query works really fast, it takes just about half a second to complete on the same large data set.